30.0.0 (#848)
MetaMask/controllers: CHANGELOG.md

## [Unreleased]

## [30.0.0]
### Uncategorized
- Introduce Detected Tokens ([#808](https://github.com/MetaMask/controllers/pull/808))
- Always pull from dynamic token list from TokenListController ([#806](https://github.com/MetaMask/controllers/pull/806))

### Added
- **BREAKING:** Expose chainId to AssetsContractController and add check whether current network supports token detection ([#809](https://github.com/MetaMask/controllers/pull/809))
  - Consumers will need to pass listener method `onNetworkStateChange` in `AssetsContractController` constructor options object. This method should be called with network/provider details when network changes occur.
- Add "onCollectibleAdded" event handler to the CollectiblesController constructor ([#814](https://github.com/MetaMask/controllers/pull/814))
  - This event handler was added to allow us to capture metrics.

> - Append phishfort blocklist to list used by PhishingController ([#715](https://github.com/MetaMask/controllers/pull/715))

Reviewer: This should be listed under "changed", and it includes breaking changes. Specifically the `test` method now returns a different result. We'll also want to include advise about updating designs to ensure we don't direct people disputing a Phishfort block to make an issue on our `eth-phishing-detect` repo (as our current designs would do if they were unaltered after this update).

Reviewer 2: added the note about breaking change here:https://github.com/MetaMask/controllers/pull/848/commits/5114337b3812caf7511456fabc63215bd7e45e73.

> We'll also want to include advice about updating designs to ensure we don't direct people disputing a Phishfort block to make an issue on our eth-phishing-detect repo

not clear on what this would look like.

Reviewer: Maybe something like:

```
  - Designs may need to be updated to account for the fact that sites may now be blocked by Phishfort. We should ensure users are not directed to the `eth-phishing-detect` repository to dispute Phishfort blocks, because we will not be able to help them.
```

Reviewer: I created a ticket for handing this on `phishing-warning` as well: https://github.com/MetaMask/phishing-warning/issues/15
cc @mcmire @danjm

Reviewer 2: add your comment here: https://github.com/MetaMask/controllers/pull/848/commits/87da72b1fae8ec3b18c866812897e725e95b44b8

## [Unreleased]

## [30.0.0]
### Added
- **BREAKING:** Introduce `getNetworkState`, `getPreferencesState`, and `onTokenListStateChange` to the TokenDetectionController constructor options object ([#808](https://github.com/MetaMask/controllers/pull/808))

>   - getPreferencesState provides the default value for `useTokenDetection` from the PreferencesController state.

Reviewer 2: updated here: https://github.com/MetaMask/controllers/pull/848/commits/8f0b66e74d974e5e1da4d250f99b289e2eff70c3

  - getNetworkState provides the default value for `chainId` from the NetworkController state.
  - onTokenListStateChange listener triggers a detect action whenever the TokenListController finishes fetching a token list.

> - **BREAKING:** Expose `chainId` to AssetsContractController and add check whether current network supports token detection ([#809](https://github.com/MetaMask/controllers/pull/809))

Reviewer 2: updated here: https://github.com/MetaMask/controllers/pull/848/commits/8f0b66e74d974e5e1da4d250f99b289e2eff70c3

  - Consumers will need to pass listener method `onNetworkStateChange` in `AssetsContractController` constructor options object. This method should be called with network/provider details when network changes occur.
- Add `onCollectibleAdded` event handler to the CollectiblesController constructor ([#814](https://github.com/MetaMask/controllers/pull/814))
  - This event handler was added to allow us to capture metrics.

### Changed

> - **BREAKING:** Change `addTokens` to `addDetectedTokens` on the TokenDetectionController constructor ([#808](https://github.com/MetaMask/controllers/pull/808))

Reviewer 3: Wording: "Rename `addTokens` on the TokenDetectionController constructor options object to `addDetectedTokens`"?

Reviewer 2: updated here: https://github.com/MetaMask/controllers/pull/848/commits/8f0b66e74d974e5e1da4d250f99b289e2eff70c3

  - We are no longer automatically adding detected tokens to the wallet. This will provide users with the ability to manually import or ignore detected tokens.
- **BREAKING:** Rename `useStaticTokenList` to `useTokenDetection` on the `PreferencesController` constructor options object and set the value to be true by default ([#808](https://github.com/MetaMask/controllers/pull/808))
  - Token detection will now be enabled by default.
- **BREAKING:** Append phishfort blocklist to list used by PhishingController ([#715](https://github.com/MetaMask/controllers/pull/715))
  - The `test` method on `PhishingController` no longer returns a boolean, it now returns an object matching the `EthPhishingDetectResult` interface (defined in `PhishingController.ts`).
  - Designs may need to be updated to account for the fact that sites may now be blocked by Phishfort. We should ensure users are not directed to the `eth-phishing-detect` repository to dispute Phishfort blocks, because we will not be able to help them.

> - **BREAKING:** Rename `convertPriceToDecimal` method name to `convertHexToDecimal` ([#808](https://github.com/MetaMask/controllers/pull/808))

Reviewer 3: This is a function, not a method.

>   - Consumers of this method must update the function name wherever it is used.

Reviewer 3: I think this is probably obvious since it's a rename, so I think you can remove this line.

Reviewer 2: updated here: https://github.com/MetaMask/controllers/pull/848/commits/8f0b66e74d974e5e1da4d250f99b289e2eff70c3

- Update TokensController to support detected tokens ([#808](https://github.com/MetaMask/controllers/pull/808))
  - Introduce `detectedTokens` to the config object to track detected tokens, which is updated by calling `addDetectedTokens` whenever TokenDetectionController detects new tokens.

>   - Added `fetchTokenMetadata` private method to fetch token metadata whenever adding individual tokens. This is currently used to populate aggregator information.

Reviewer 3: This isn't private, nor is it a method. Maybe say "Added `fetchTokenMetadata` utility function to...."?

Reviewer 2: updated here: https://github.com/MetaMask/controllers/pull/848/commits/8f0b66e74d974e5e1da4d250f99b289e2eff70c3

Reviewer 4: Probably not a good idea on my part to name the method within the class to be the same as the utility function that it's importing. There is actually a private method that lives in the TokensController controller named `fetchTokenMetadata`. :sweat_smile:

Reviewer 3: Ah, okay!

- Append `detectedTokens` to both TokenBalancesController & TokenRatesController `tokens` config object within their `onTokensStateChange` listener methods ([#808](https://github.com/MetaMask/controllers/pull/808))
  - This change ensures that we are populating both balances and rates for detected tokens.
- Re-point OpeanSea API calls to Codefi proxy ([#805](https://github.com/MetaMask/controllers/pull/805))
  - All NFT related queries that were routed to OpenSea will now first route to a proxy server owened by Codefi. If this first request fails, and an OpenSea API key has been set, the query will re-route to OpenSea as a fallback.

> - Return ETH currency rate for testnet ETH ([#816](https://github.com/MetaMask/controllers/pull/816))

Reviewer 2: Yes I can re-word this. Good idea.

> - Return ETH currency rate for testnet ETH ([#816](https://github.com/MetaMask/controllers/pull/816))

Reviewer 2: updated here: https://github.com/MetaMask/controllers/pull/848/commits/8f0b66e74d974e5e1da4d250f99b289e2eff70c3

- Query for name and symbol on ERC721 contracts regardless of whether or not they support the metadata interface ([#834](https://github.com/MetaMask/controllers/pull/834))

> - Increase polling interval for TokenListController and require minimum of 3 occurences across sources for use in the dynamic token list ([#836](https://github.com/MetaMask/controllers/pull/836))

Reviewer 3: Typo: occurences -> occurrences.

Also what does "require minimum of 3 occurrences" mean? That the token must occur in 3 public token lists in order for us to autodetect it? Sorry I'm not caught up on this.

Reviewer 2: improved clarity here: https://github.com/MetaMask/controllers/pull/848/commits/8f0b66e74d974e5e1da4d250f99b289e2eff70c3

Reviewer 3: Thanks!

### Removed

> - **BREAKING:** Deprecate `removeAndIgnoreToken` method in TokensController ([#808](https://github.com/MetaMask/controllers/pull/808))

Reviewer 3: FYI: We want to be careful about wording here. If we deprecate something, it means that we are warning people that we are _about_ to remove that thing in the next release, but we haven't quite taken it away yet. We appear to be removing this method outright, though, so what do we think about saying "Remove `removeAndIgnoreToken` from TokensController"?

Reviewer 2: updated here: https://github.com/MetaMask/controllers/pull/848/commits/8f0b66e74d974e5e1da4d250f99b289e2eff70c3

>   - The logic for removing token(s) is now consolidated in a new method named `ignoredTokens`. Consumers will need to update instances of removeAndIgnoreToken to use ignoredTokens instead.

Reviewer 3: Nit: Use backticks around "removeAndIgnoreToken" and "ignoredTokens"?

Reviewer 2: updated here: https://github.com/MetaMask/controllers/pull/848/commits/8f0b66e74d974e5e1da4d250f99b289e2eff70c3

- **BREAKING:** Remove `onTokensStateChange` from the TokenDetectionController constructor options object ([#808](https://github.com/MetaMask/controllers/pull/808))
  - This was previously used to update the `tokens` property in the controller's configuration object. This is not needed since those tokens are managed by the TokensController.

> - **BREAKING:** Remove `useStaticTokenList` and `onPreferencesStateChange` from TokenListController constructor options object ([#806](https://github.com/MetaMask/controllers/pull/806))

Reviewer 2: cc @Cal-L

  - useStaticTokenList was previously used to determined if this controller fetched against a static vs dynamic token list and onPreferencesStateChange was used to update useStaticTokenList.
  - The controller now always fetches from a dynamic token list.
- **BREAKING:** Remove snap-specific network-access endowment ([#820](https://github.com/MetaMask/controllers/pull/820))

>   - Consumers who still require this endowment should import from @metamask/snap-controllers minimum version 0.13.0

Reviewer 2: updated here: https://github.com/MetaMask/controllers/pull/848/commits/8f0b66e74d974e5e1da4d250f99b289e2eff70c3

### Fixed

> - Fix issues parsing non-standard ERC20 responses ([#830](https://github.com/MetaMask/controllers/pull/830))

Reviewer 2: updated here: https://github.com/MetaMask/controllers/pull/848/commits/8f0b66e74d974e5e1da4d250f99b289e2eff70c3

>  - Now correctly reads token contract values for decimals and symbol with solidity return type `bytes32`

Reviewer 2: fixed here: https://github.com/MetaMask/controllers/pull/848/commits/8f0b66e74d974e5e1da4d250f99b289e2eff70c3

  - This event handler was added to allow us to capture metrics.

### Changed
- **BREAKING:** Change `addTokens` to `addDetectedTokens` on the TokenDetectionController constructor ([#808](https://github.com/MetaMask/controllers/pull/808))
  - We are no longer automatically adding detected tokens to the wallet. This will provide users with the ability to manually import or ignore detected tokens.
- **BREAKING:** Rename `useStaticTokenList` to `useTokenDetection` on the `PreferencesController` constructor options object and set the value to be true by default ([#808](https://github.com/MetaMask/controllers/pull/808))
  - Token detection will now be enabled by default.
- **BREAKING:** Append phishfort blocklist to list used by PhishingController ([#715](https://github.com/MetaMask/controllers/pull/715))
  - The `test` method on `PhishingController` no longer returns a boolean, it now returns an object matching the `EthPhishingDetectResult` interface (defined in `PhishingController.ts`).
- **BREAKING:** Rename `convertPriceToDecimal` method name to `convertHexToDecimal` ([#808](https://github.com/MetaMask/controllers/pull/808))

>   - Consumers of the controllers repo may have to update the instances where this function is used.

Reviewer 2: ```suggestion
  - Consumers of this method must update the function name wherever it is used.
```
